File: examples/loss.rs
//! A real training log: per-step minibatch loss of topos's makemore bigram
//! model (see examples/data/README.md), its rolling mean, and the corpus's known
//! bigram limit as a target rule. No synthetic data — this training actually ran.

use malevich::{Dash, Frame, Line, Plot, Rule};

fn main() {
    let (steps, losses): (Vec<f64>, Vec<f64>) = include_str!("data/topos_loss.csv")
        .lines()
        .filter_map(|line| {
            let (step, loss) = line.split_once(',')?;
            let step: f64 = step.parse().ok()?;
            let loss: f64 = loss.parse().ok()?;
            Some((step, loss))
        })
        .unzip();
    let smoothed = malevich::stat::Window::new(25).mean(&losses);

    let plot = Plot::new()
        .layer(Line::xy(&steps[..], &losses[..]).label("minibatch"))
        .layer(
            Line::xy(&steps[..], &smoothed[..])
                .label("rolling mean")
                .glow(),
        )
        .layer(Rule::h(2.45).label("bigram limit").dash(Dash::Dashed))
        .title("topos: bigram training on 32k names")
        .x_label("step")
        .y_label("loss");
    println!("{}", plot.render_best(&Frame::plain(76, 19)));
}

File: js/examples/loss.ts
/** Training-loop look: two series, a target rule, an annotation. */

import { Frame, Line, Plot, Rule, Text } from "../dist/index.js";

const steps = Float64Array.from({ length: 120 }, (_, i) => i);
const train = Float64Array.from(
  steps,
  (s) => 3.8 * Math.exp(-0.035 * s) + 0.32 + 0.05 * Math.sin(s * 0.7),
);
const val = Float64Array.from(
  steps,
  (s) => 4.0 * Math.exp(-0.03 * s) + 0.55 + 0.08 * Math.cos(s * 0.35),
);

console.log(
  new Plot()
    .layer(Line.xy(steps, train).label("train"))
    .layer(Line.xy(steps, val).label("val"))
    .layer(Rule.h(0.5).label("target"))
    .layer(Text.at(60, 2, "< converging"))
    .title("loss (synthetic)")
    .xLabel("step")
    .yLabel("loss")
    .render(Frame.detect()),
);
